### ISSUE TITLE
Stupid example for using declarations

### ISSUE THREAD

**Issue by Haukinger** [dotnet/docs#20295]
The example says that

    using var x = new SomeDisposable();
    var y = 0;
    y += x.Stuff();
    return y;

is equivalent to

    var y = 0;
    using (var x = new SomeDisposable())
    {
         y += x.Stuff();
    }
    return y;
 
while in fact it's as equivalent to

    using (var x = new SomeDisposable())
    {
         var y = 0;
         y += x.Stuff();
         return y;
    }

I'd prefer an example that shows more benefit than rationalizing a pair of parentheses and a pair of braces.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c36588d0-6bdf-1740-220c-10b20bb88730
* Version Independent ID: 587ed47f-f532-e018-9b0d-6ee76afd20e4
* Content: [What's new in C# 8.0 - C# Guide](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8)
* Content Source: [docs/csharp/whats-new/csharp-8.md](https://github.com/dotnet/docs/blob/master/docs/csharp/whats-new/csharp-8.md)
* Product: **dotnet-csharp**
* Technology: **csharp-whats-new**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by BillWagner** [comment]
Thanks for writing this @Haukinger 

In the "what's new" section, we try to keep the examples small and focused. That keeps these articles to a reasonable length, while giving a brief explanation for each feature. The language reference and tutorial pages have more real-world examples.

That said, this could be better. I've added it to our backlog to address when we next update this article.

**Comment by mohsin-mehmood** [comment]
@BillWagner this [pull](https://github.com/dotnet/docs/pull/20411) request is merged and should fix this issue

**Comment by bnewcomb** [comment]
@Haukinger 

The benefit of the **using declaration** is that you don't need to worry about nested scope and variable lifetime, but you still benefit from not needing to explicitly call Dispose() before you return from your method.

The examples are demonstrating how your choice between a **using statement** and a **using declaration** would depend on how soon you'd like to Dispose the object; before the end of the method or with the end of the method.    The using declaration would dispose of the file once it goes out of the enclosing scope, be it method scope or block scope.  This means you could also declare it inside a for loop and it would be disposed after the closing brace of the loop and before the return statement.

In the example of the **using declaration**, if you had additional work to do prior to the return statement, the object wouldn't be disposed until you call return.  The **using declaration** method is easier for simple cases because it doesn't create a nested block level scope inside the method so you don't have to be mindful of the scope of any variables as the return statement and the variable declaration it depends on are at the same scope.  However, for example, if you popped up a message box on the desktop after the **using declaration** but before the return statement, you could keep a file locked on a network drive preventing others from being able to access it until execution continued past the return statement.  Or you could literally run out of RAM depending on how big the file is and how much processing you are doing.

In contrast, if you used a **using statement** or you nested the **using declaration** inside a for loop for example, you could free up the file before the method returns as Dispose() is called when execution passes the closing brace of the **using statement** or the closing brace of the for loop.  If you popped up your message box after the brace, the network file would be available to others even if you left your desk for coffee and your code is sitting there waiting on you.  The **using statement** creates a nested scope inside of the method scope, called block level scope, that can impact when the object is disposed as well as the lifetime of variables declared inside that scope.  If skippedLines "falls out of scope" because the declaration occurred inside the for loop or inside the using statement, it won't be available to be returned at the end of your method.  It's just like trying to access a variable that is declared inside another method you just called after you called it and returned from it.  The variable no longer exists in scope and can't be read.

In your example, the method would return before the closing brace of the **using statement**.  As long as you didn't have more code prior to the return statement, it's functionally equivalent, but if you inserted more code the object would remain in memory or your file would remain locked until execution reached the return statement.

**Comment by bnewcomb** [comment]
> @bnewcomb I was hoping that there's actually more to the using declaration than just skipping a pair of braces, but you're reinforcing by understanding. So the using declaration just says "I don't care _when_ this is disposed, as long as it's disposed _at all_."

More like "I want this to be disposed once all the operations that follow the declaration have completed and it is no longer needed" 

In the example, that is once this method finishes and control returns to the method that called it. 

We need a good metaphor. The Super Bowl Party. 

You can bring a plate and leave with whatever it can carry. 

Using Declaration - You can have as much food as you want during the game. Take as many trips to the kitchen table as you like. How much  you have to drink(processor) , the size of your stomach(HD) and plates(ram), Number of people in attendance(throughput). All factors. After the game, food is immediately thrown into the trash so you better have a plate made up for your family at home. If you stay out too late, you'll have some unhappy people when you get back. 

Using Statement - Caterers come at the start of the 2nd quarter.  You can make a plate, but then everything is thrown away to make room for the halftime show. Sorry. Food or Cheerleaders. Not enough room for both. You can create plates to store food for later and to share with the cheerleaders. It's considerate and may impact what you end up taking back home.(story, phone number, etc) 

Instead of needing to perform the operations and variable assignment(getting a plate) inside the braces(2nd quarter), you can start after the declaration(coin toss) and eat all through the game. Less fussing with plates and moving food around. The smokies are best warm, right from the crock pot. Dispose(food tossed) will get called at the end(post game analysis) before control returns to the caller(take plate home and give to wife). 

If you didn't care at all, it could wait until you stop the app. In the case of a web app, that could be a very long time. In the case of the Super Bowl, the food would be rotten and house unlivable. Think fraternity house with no rushes. Things can get very disgusting before the health department steps in. (Garbage Collector)

What kind of party do you want to go to?

**Comment by BillWagner** [comment]
Thanks for all the discussion and the PR to fix the original issue.

I explain the difference between the **using declaration** and the **using statement** like this:

- In most cases, you'll close the `using` statement at the end of the enclosing block. In all those cases, the `using` declaration saves a couple brackets and parens. (It is that simple). It's most useful when a method reads from one resource and writes to another. Both resources should be disposed at the end of the enclosing block, so it's double the savings.
- Prefer the `using` statement when you want to dispose of the resource *before* the end of the enclosing block.

As the original issue was fixed, I'm going to close this issue.

**Comment by Haukinger** [comment]
@BillWagner double savings only, if the using statements cannot immediately follow each other, right?

    using( var readingStream = new ...)
    using( var writingStream = new ...)
    {
        // ...
    }

and

    using( var readingStream = new ...)
    {
        // ...read something...
        using( var writingStream = new ...)
        {
            // ...
        }
    }
